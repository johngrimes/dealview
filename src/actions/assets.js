// @flow

import { putObject, deleteObject, getAllObjects } from 'db/db'
import { invalidateBalanceSheet } from 'actions/balanceSheet'
import type { Asset, AssetWithId, AssetMap } from 'types/assets/asset'
import type { Thunk } from 'types/commonTypes'

const objectStore = 'Asset'

type PutAssetRequestAction = {
  type: 'PUT_ASSET_REQUEST',
  asset: Asset,
}
type PutAssetSuccessAction = {
  type: 'PUT_ASSET_SUCCESS',
  asset: AssetWithId
}
type PutAssetFailureAction = {
  type: 'PUT_ASSET_FAILURE',
  error: string|null
}

type DeleteAssetRequestAction = {
  type: 'DELETE_ASSET_REQUEST',
  id: string,
}
type DeleteAssetSuccessAction = {
  type: 'DELETE_ASSET_SUCCESS',
  id: string
}
type DeleteAssetFailureAction = {
  type: 'DELETE_ASSET_FAILURE',
  error: string|null
}

type LoadAssetsRequestAction = {
  type: 'LOAD_ASSETS_REQUEST'
}
type LoadAssetsSuccessAction = {
  type: 'LOAD_ASSETS_SUCCESS',
  assets: AssetMap
}
type LoadAssetsFailureAction = {
  type: 'LOAD_ASSETS_FAILURE',
  error: string|null
}

export type AssetAction = PutAssetRequestAction
                        | PutAssetSuccessAction
                        | PutAssetFailureAction
                        | DeleteAssetRequestAction
                        | DeleteAssetRequestAction
                        | DeleteAssetFailureAction
                        | LoadAssetsRequestAction
                        | LoadAssetsSuccessAction
                        | LoadAssetsFailureAction

export const putAssetRequest = (asset: Asset): PutAssetRequestAction => {
  return {
    type: 'PUT_ASSET_REQUEST',
    asset,
  }
}

export const putAssetSuccess = (asset: AssetWithId): PutAssetSuccessAction => {
  return {
    type: 'PUT_ASSET_SUCCESS',
    asset,
  }
}

export const putAssetFailure = (error: string|null): PutAssetFailureAction => {
  return {
    type: 'PUT_ASSET_FAILURE',
    error,
  }
}

export const deleteAssetRequest = (id: string): DeleteAssetRequestAction => {
  return {
    type: 'DELETE_ASSET_REQUEST',
    id,
  }
}

export const deleteAssetSuccess = (id: string): DeleteAssetSuccessAction => {
  return {
    type: 'DELETE_ASSET_SUCCESS',
    id,
  }
}

export const deleteAssetFailure = (error: string|null): DeleteAssetFailureAction => {
  return {
    type: 'DELETE_ASSET_FAILURE',
    error,
  }
}

export const loadAssetsRequest = (): LoadAssetsRequestAction => {
  return { type: 'LOAD_ASSETS_REQUEST' }
}

export const loadAssetsSuccess = (assets: AssetMap): LoadAssetsSuccessAction => {
  return {
    type: 'LOAD_ASSETS_SUCCESS',
    assets,
  }
}

export const loadAssetsFailure = (error: string|null): LoadAssetsFailureAction => {
  return {
    type: 'LOAD_ASSETS_FAILURE',
    error,
  }
}

export const putAsset = (asset: Asset): Thunk => {
  return dispatch => {
    dispatch(putAssetRequest(asset))
    return new Promise(async (resolve, reject) => {
      try {
        const saved = await putObject(objectStore, asset)
        dispatch(invalidateBalanceSheet())
        dispatch(putAssetSuccess(saved))
        resolve(saved)
      } catch (error) {
        dispatch(putAssetFailure(error))
        reject(error)
      }
    })
  }
}

export const deleteAsset = (id: string): Thunk => {
  return dispatch => {
    dispatch(deleteAssetRequest(id))
    return new Promise(async (resolve, reject) => {
      try {
        const key = await deleteObject(objectStore, id)
        dispatch(invalidateBalanceSheet())
        dispatch(deleteAssetSuccess(key))
        resolve(key)
      } catch (error) {
        dispatch(deleteAssetFailure(error))
        reject(error)
      }
    })
  }
}

export const loadAssets = (): Thunk => {
  return dispatch => {
    dispatch(loadAssetsRequest())
    return new Promise(async (resolve, reject) => {
      try {
        const assets = await getAllObjects(objectStore)
        dispatch(loadAssetsSuccess(assets))
        resolve(assets)
      } catch (error) {
        dispatch(loadAssetsFailure(error))
        reject(error)
      }
    })
  }
}
