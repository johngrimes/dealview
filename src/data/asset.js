import moment from 'moment'

export const getValuationAtDate = (asset, date) => {
  const now = moment().startOf('day')
  const startDate = asset.purchaseDate || now
  if (
    date.isBefore(startDate) ||
    (asset.saleDate && date.isAfter(asset.saleDate))
  ) {
    return 0
  }
  const daysSinceStart = Math.floor(
    moment.duration(date.diff(startDate)).asDays()
  )
  const startValue = asset.purchaseAmount
    ? parseInt(asset.purchaseAmount, 10)
    : parseInt(asset.currentValue, 10)
  const forecastGrowth = parseInt(asset.forecastGrowth, 10) / 10000 / 365
  return Math.round(startValue * Math.pow(1 + forecastGrowth, daysSinceStart))
}
