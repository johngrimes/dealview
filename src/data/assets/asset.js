// @flow

export type Asset = {
  type: 'RealEstate',
  instanceId: string,
  name: string,
  lastValuation?: number
}

export type AssetWithId = Asset & { id: string }

export type AssetMap = { [id: string]: AssetWithId }
