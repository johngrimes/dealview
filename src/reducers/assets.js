// @flow

import _ from 'lodash'
import type { Reducer, State, Action } from 'redux'

import type { AssetMap } from 'data/assets/asset'
import type { AssetAction } from 'actions/assets'

export type AssetState = {
  status: 'uninitialised'|'loading'|'loaded'|'error',
  error?: string,
  objects: AssetMap
}
const initialState: AssetState = {
  status: 'uninitialised',
  objects: {},
}

const AssetsReducer: Reducer<State, Action> = (state = initialState, action: AssetAction) => {
  switch (action.type) {
    case 'PUT_ASSET_REQUEST':
    case 'DELETE_ASSET_REQUEST':
    case 'LOAD_ASSETS_REQUEST':
      return {
        ...state,
        status: 'loading',
      }
    case 'PUT_ASSET_SUCCESS':
      return {
        ...state,
        status: 'loaded',
        objects: { ...state.objects, [action.asset.id]: action.asset },
      }
    case 'DELETE_ASSET_SUCCESS':
      return {
        ...state,
        status: 'loaded',
        objects: _.omit(state.objects, action.id),
      }
    case 'LOAD_ASSETS_SUCCESS':
      return {
        ...state,
        status: 'loaded',
        objects: action.assets,
      }
    case 'PUT_ASSET_FAILURE':
    case 'DELETE_ASSET_FAILURE':
    case 'LOAD_ASSETS_FAILURE':
      return {
        ...state,
        status: 'error',
        error: action.error,
      }
    default:
      return state
  }
}

export default AssetsReducer
