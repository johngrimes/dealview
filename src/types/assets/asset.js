// @flow

import moment from 'moment'

import { DateFormat } from 'types/commonTypes'
import type { Valuations, Valuation } from 'components/forms/ValuationsInput'

export type Asset = {
  type: 'RealEstate',
  instanceId: string,
  name: string,
  startDate: string,
  endDate?: string,
  valuations: Valuations
}

export type AssetWithId = Asset & { id: string }

export type AssetMap = { [id: string]: AssetWithId }

export const getValuationAtDate = (asset: { startDate: string, endDate?: string, valuations: Valuations },
                                   date: string): number => {
  const queryDate = moment(date)
  const startDate = moment(asset.startDate)
  const endDate = typeof asset.endDate === 'string'
    ? moment(asset.endDate)
    : undefined
  if (typeof asset.valuations === 'undefined' ||
      asset.valuations.length === 0 ||
      queryDate.isBefore(startDate) ||
      (endDate && queryDate.isAfter(endDate))) {
    return 0
  }
  const sortedValuations = asset.valuations.slice().sort(compareValuationsByDate)
  const idxFuture = sortedValuations.findIndex(v => {
    return moment(v.date, DateFormat).valueOf() > queryDate.valueOf()
  })
  const slicedValuations = idxFuture === -1
    ? sortedValuations
    : sortedValuations.slice(0, idxFuture)
  const amount = typeof slicedValuations[slicedValuations.length - 1] === 'undefined'
    ? 0
    : slicedValuations[slicedValuations.length - 1].amount
  return typeof amount === 'number' ? amount : 0
}

const compareValuationsByDate = (a: Valuation, b: Valuation): number => {
  const [ milliA, milliB ] = [ a, b ].map(v => { return v.date ? moment(v.date).valueOf() : 0 })
  return milliA - milliB
}
