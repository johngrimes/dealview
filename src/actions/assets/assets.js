// @flow

import { putObject, deleteObject, getAllObjects } from 'db/db'
import { invalidateBalanceSheet } from 'actions/balanceSheet'
import { createObjectActions } from 'actions/objects'
import type { Asset } from 'types/assets/asset'
import type { Thunk } from 'types/commonTypes'

const objectStore = 'Asset'

const actions = createObjectActions('ASSET', 'ASSETS',
                                    'Asset', 'Assets')

const putAsset = (asset: Asset): Thunk => {
  return dispatch => {
    dispatch(actions.putAssetRequest(asset))
    return new Promise(async (resolve, reject) => {
      try {
        const saved = await putObject(objectStore, asset)
        dispatch(invalidateBalanceSheet())
        dispatch(actions.putAssetSuccess(saved))
        resolve(saved)
      } catch (error) {
        dispatch(actions.putAssetFailure(error))
        reject(error)
      }
    })
  }
}

const deleteAsset = (id: string): Thunk => {
  return dispatch => {
    dispatch(actions.deleteAssetRequest(id))
    return new Promise(async (resolve, reject) => {
      try {
        const key = await deleteObject(objectStore, id)
        dispatch(invalidateBalanceSheet())
        dispatch(actions.deleteAssetSuccess(key))
        resolve(key)
      } catch (error) {
        dispatch(actions.deleteAssetFailure(error))
        reject(error)
      }
    })
  }
}

const loadAssets = (): Thunk => {
  return dispatch => {
    dispatch(actions.loadAssetsRequest())
    return new Promise(async (resolve, reject) => {
      try {
        const assets = await getAllObjects(objectStore)
        dispatch(actions.loadAssetsSuccess(assets))
        resolve(assets)
      } catch (error) {
        dispatch(actions.loadAssetsFailure(error))
        reject(error)
      }
    })
  }
}

export default {
  ...actions,
  putAsset,
  deleteAsset,
  loadAssets,
}
