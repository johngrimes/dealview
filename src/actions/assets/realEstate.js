// @flow

import { putObject, deleteObject, getAllObjects } from 'db/db'
import AssetActions from 'actions/assets/assets'
import { realEstateToAsset } from 'types/assets/realEstate'
import { createObjectActions } from 'actions/objects'
import type { RealEstate } from 'types/assets/realEstate'
import type { Thunk } from 'types/commonTypes'

const objectStore = 'Asset.RealEstate'

const actions = createObjectActions('REAL_ESTATE', 'REAL_ESTATE',
                                    'RealEstate', 'RealEstate')

const putRealEstate = (realEstate: RealEstate): Thunk => {
  return dispatch => {
    dispatch(actions.putRealEstateRequest(realEstate))
    return new Promise(async (resolve, reject) => {
      try {
        const saved = await putObject(objectStore, realEstate)
        dispatch(actions.putRealEstateSuccess(saved))
        await dispatch(AssetActions.putAsset(realEstateToAsset(saved)))
        resolve(saved)
      } catch (error) {
        dispatch(actions.putRealEstateFailure(error))
        reject(error)
      }
    })
  }
}

const deleteRealEstate = (id: string): Thunk => {
  return dispatch => {
    dispatch(actions.deleteRealEstateRequest(id))
    return new Promise(async (resolve, reject) => {
      try {
        const deletedId = await deleteObject(objectStore, id)
        dispatch(AssetActions.deleteAsset(deletedId))
        dispatch(actions.deleteRealEstateSuccess(deletedId))
        resolve(deletedId)
      } catch (error) {
        dispatch(actions.deleteRealEstateFailure(error))
        reject(error)
      }
    })
  }
}

const loadRealEstate = (): Thunk => {
  return dispatch => {
    dispatch(actions.loadRealEstateRequest())
    return new Promise((resolve, reject) => {
      getAllObjects(objectStore).then(assets => {
        dispatch(actions.loadRealEstateSuccess(assets))
        resolve(assets)
      }).catch(error => {
        dispatch(actions.loadRealEstateFailure(error))
        reject(error)
      })
    })
  }
}

export default {
  ...actions,
  putRealEstate,
  deleteRealEstate,
  loadRealEstate,
}
