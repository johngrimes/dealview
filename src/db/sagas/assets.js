// @flow

import { take, takeEvery, fork, call, put, all } from 'redux-saga/effects'

import * as AssetActions from 'actions/realEstate'
import { putObject, deleteObject, getAllObjects } from 'db/db'
import type { Asset } from 'types/assets/asset'
import type { DeleteAssetRequestAction } from 'assets/actions'

const objectStore = 'Asset'

function* watchPutAsset() {
  while (true) {
    const assetAction = ('PUT_ASSET_REQUEST')
    yield fork(putAsset, assetAction.asset)
    const realEstateAction = yield take('PUT_REAL_ESTATE_REQUEST')
    yield fork(putAsset, realEstateAction.realEstate)
  }
}

function* putAsset(asset: Asset) {
  try {
    const saved = yield call(putObject, objectStore, asset)
    yield put(AssetActions.PutAssetSuccessAction(saved))
  } catch (error) {
    yield put(AssetActions.PutAssetFailureAction(error))
  }
}

function* watchDeleteAsset() {
  yield takeEvery('DELETE_ASSET_REQUEST', deleteAsset)
}

function* deleteAsset(action: DeleteAssetRequestAction) {
  try {
    const id = yield call(deleteObject, objectStore, action.id)
    yield put(AssetActions.DeleteAssetSuccessAction(id))
  } catch (error) {
    yield put(AssetActions.DeleteAssetFailureAction(error))
  }
}

function* watchLoadAsset() {
  yield takeEvery('LOAD_ASSET_REQUEST', loadAsset)
}

function* loadAsset() {
  try {
    const realEstate = yield call(getAllObjects, objectStore)
    yield put(AssetActions.LoadAssetSuccessAction(realEstate))
  } catch (error) {
    yield put(AssetActions.LoadAssetFailureAction(error))
  }
}

export function* watchAssets() {
  yield all([
    watchPutAsset(),
    watchDeleteAsset(),
    watchLoadAsset(),
  ])
}
