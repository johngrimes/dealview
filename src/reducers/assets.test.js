/* global expect */

import AssetsReducer from './assets.js'
import * as AssetActions from '../actions/assets.js'
import { validAssetWithId1, validAssetWithId2 } from '../test/fixtures/asset.js'

describe('AssetsReducer', () => {
  const initialState = {
    status: 'loaded',
    objects: {
      [validAssetWithId1.id]: validAssetWithId1
    }
  }

  it('should update state upon request', () => {
    const actions = [
      AssetActions.putAssetRequest,
      AssetActions.deleteAssetRequest,
      AssetActions.loadAssetRequest
    ]
    actions.forEach(action => {
      const nextState = AssetsReducer(initialState, action())
      expect(nextState).toMatchSnapshot()
    })
  })

  it('should update state upon put success', () => {
    const action = AssetActions.putAssetSuccess(validAssetWithId2)
    const nextState = AssetsReducer(initialState, action)
    expect(nextState).toMatchSnapshot()
  })

  it('should update state upon delete success', () => {
    const action = AssetActions.deleteAssetSuccess(validAssetWithId1.id)
    const nextState = AssetsReducer(initialState, action)
    expect(nextState).toMatchSnapshot()
  })

  it('should update state upon load success', () => {
    const action = AssetActions.loadAssetSuccess({
      [validAssetWithId1.id]: validAssetWithId1,
      [validAssetWithId2.id]: validAssetWithId2
    })
    const nextState = AssetsReducer(initialState, action)
    expect(nextState).toMatchSnapshot()
  })

  it('should update state upon failure', () => {
    const actions = [
      AssetActions.putAssetFailure,
      AssetActions.deleteAssetFailure,
      AssetActions.loadAssetFailure
    ]
    actions.forEach(action => {
      const nextState = AssetsReducer(initialState, action('Some error'))
      expect(nextState).toMatchSnapshot()
    })
  })
})
