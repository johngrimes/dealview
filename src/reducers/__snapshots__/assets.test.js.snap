// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssetsReducer should update state upon delete success 1`] = `
Object {
  "objects": Object {},
  "status": "loaded",
}
`;

exports[`AssetsReducer should update state upon failure 1`] = `
Object {
  "error": "Some error",
  "objects": Object {
    "98": Object {
      "id": "98",
      "instanceId": "54",
      "lastValuation": 705000,
      "name": "5 Sunrise St",
      "type": "RealEstate",
    },
  },
  "status": "error",
}
`;

exports[`AssetsReducer should update state upon failure 2`] = `
Object {
  "error": "Some error",
  "objects": Object {
    "98": Object {
      "id": "98",
      "instanceId": "54",
      "lastValuation": 705000,
      "name": "5 Sunrise St",
      "type": "RealEstate",
    },
  },
  "status": "error",
}
`;

exports[`AssetsReducer should update state upon failure 3`] = `
Object {
  "error": "Some error",
  "objects": Object {
    "98": Object {
      "id": "98",
      "instanceId": "54",
      "lastValuation": 705000,
      "name": "5 Sunrise St",
      "type": "RealEstate",
    },
  },
  "status": "error",
}
`;

exports[`AssetsReducer should update state upon load success 1`] = `
Object {
  "objects": Object {
    "98": Object {
      "id": "98",
      "instanceId": "54",
      "lastValuation": 705000,
      "name": "5 Sunrise St",
      "type": "RealEstate",
    },
    "99": Object {
      "id": "99",
      "instanceId": "34",
      "lastValuation": 260000,
      "name": "38/8 Briggs Road",
      "type": "RealEstate",
    },
  },
  "status": "loaded",
}
`;

exports[`AssetsReducer should update state upon put success 1`] = `
Object {
  "objects": Object {
    "98": Object {
      "id": "98",
      "instanceId": "54",
      "lastValuation": 705000,
      "name": "5 Sunrise St",
      "type": "RealEstate",
    },
    "99": Object {
      "id": "99",
      "instanceId": "34",
      "lastValuation": 260000,
      "name": "38/8 Briggs Road",
      "type": "RealEstate",
    },
  },
  "status": "loaded",
}
`;

exports[`AssetsReducer should update state upon request 1`] = `
Object {
  "objects": Object {
    "98": Object {
      "id": "98",
      "instanceId": "54",
      "lastValuation": 705000,
      "name": "5 Sunrise St",
      "type": "RealEstate",
    },
  },
  "status": "loading",
}
`;

exports[`AssetsReducer should update state upon request 2`] = `
Object {
  "objects": Object {
    "98": Object {
      "id": "98",
      "instanceId": "54",
      "lastValuation": 705000,
      "name": "5 Sunrise St",
      "type": "RealEstate",
    },
  },
  "status": "loading",
}
`;

exports[`AssetsReducer should update state upon request 3`] = `
Object {
  "objects": Object {
    "98": Object {
      "id": "98",
      "instanceId": "54",
      "lastValuation": 705000,
      "name": "5 Sunrise St",
      "type": "RealEstate",
    },
  },
  "status": "loading",
}
`;
